# R script for quick project setup!

print("LOADING: SETUP.R")

create_dir <- function(fp){
  setwd(fp)
  print(paste("Working directory set to ", fp, sep=""))
  print("Creating necessary folders")
  dir.create(path = "Input")
  dir.create(path = "Input/Drugs")
  dir.create(path = "Input/HuRI_PPI")
  dir.create(path = "Input/literature_PPI")
  dir.create(path = "Input/DCA_adjusted_matrix")
  dir.create(path = "Input/GWAScat")
  dir.create(path = "Input/OMIM")
  dir.create(path = "Input/NCBI_annotation")
  dir.create(path = "Input/Human-mouse_homologs")
  dir.create(path = "Input/Bulk_data")
  dir.create(path = "Input/OMIM")
  dir.create(path = "Input/PanglaoDB")
  print("Creating output directories")
  dir.create(path = "Output")
  dir.create(path = "Output/Seurat_clusters")
  dir.create(path = "Output/DCA_MAST_DEGs")
  dir.create(path = "Output/DCA_MAST_DEGs_predictions")
  dir.create(path = "Output/Correlation_genes")
  dir.create(path = "Output/Correlation_genes_predictions")
  dir.create(path = "Output/Onedimensional cut-offs")
  dir.create(path = "Output/Multidimensional cut-offs")
  dir.create(path = "Output/GWAS")
  dir.create(path = "Output/OMIM")
  dir.create(path = "Output/Bulk_data")
  dir.create(path = "Output/Random_LASSO")
  dir.create(path = "Output/Other_predictions")
  dir.create(path = "Output/Cell_typing")
  print("")
  print("Place the appropriate data in the Input folder BEFORE PROCEEDING")
  print("E.g.: Place DCA output (latent.tsv & mean.tsv) in: /Input/DCA_adjusted_matrix")
  print("E.g.: Place OMIM genes in: /Input/OMIM")
}

setup_envir <- function(){
  print("INSTALL package Seurat")
  # Complicated process as some packages colided on my computer, though this is in theory replacable by install.packages("Seurat")
  install.packages("devtools")
  library(devtools)
  devtools::install_version(package = 'plotrix', version = package_version('3.6'), dependencies = T)
  library(plotrix)
  devtools::install_version(package = 'metap', version = package_version('1.1'), dependencies = T) # update none
  library(metap)
  install.packages("https://cran.r-project.org/src/contrib/Archive/mvtnorm/mvtnorm_1.0-8.tar.gz", repos=NULL, dependencies = T)
  library(mvtnorm)
  if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
  BiocManager::install("monocle")
  library(monocle)
  BiocManager::install("SingleCellExperiment")
  library(SingleCellExperiment)
  BiocManager::install("rtracklayer")
  library(rtracklayer)
  #devtools::install_version(package = 'metap', version = package_version('1.1'), dependencies = T)
  #library(metap)
  devtools::install_version(package = 'Seurat', version = package_version('3.1.0'), dependencies = F)
  library(Seurat)
  print("Seurat working!")
  
  print("INSTALL doParallel")
  install.packages("doParallel")
  library(doParallel)
  print("doParallel working!")
  
  print("INSTALL dplyr")
  install.packages("dplyr")
  library(dplyr)
  print("dplyr working!")
  
  print("INSTALL stats")
  install.packages("stats")
  library(stats)
  print("dplyr working!")
  
   print("INSTALL limma")
  install.packages("limma")
  library(limma)
  print("limma working!")
  
  print("INSTALL GEOquery")
  install.packages("GEOquery")
  library(GEOquery)
  print("GEOquery working!")
  
  print("INSTALL igraph")
  install.packages("igraph")
  library(igraph)
  print("igraph working!")
  
  print("INSTALL gespeR")
  install.packages("gespeR")
  library(gespeR)
  print("gespeR working!")
  
  print("INSTALL KEGGREST & KEGGgraph")
  BiocManager::install("KEGGREST", version = "3.6")
  library(KEGGREST)
  BiocManager::install("KEGGgraph", version = "3.6")
  library(KEGGgraph)
  print("KEGGREST & KEGGgraph working!")
  
  print("INSTALL readxl")
  install.packages("readxl")
  library(readxl)
  print("readxl working!")
  
  print("INSTALL jaccard")
  install.packages("jaccard")
  library(jaccard)
  print("jaccard working!")
  
  print("INSTALL corrplot")
  install.packages("corrplot")
  library(corrplot)
  print("corrplot working!")
  
  print("INSTALL CINNA")
  install.packages("CINNA")
  library(CINNA)
  print("CINNA working!")
  
  print("INSTALL igraph")
  install.packages("igraph")
  library(igraph)
  print("igraph working!")
  
  print("INSTALL hrbrthemes")
  install.packages("hrbrthemes")
  library(hrbrthemes)
  print("hrbrthemes working!")
  
  print("Unloading all installed packages")
  lapply(paste('package:',names(sessionInfo()$otherPkgs),sep=""),detach,character.only=TRUE,unload=TRUE)
}

fp <- paste(getwd(),"/..",sep="")
create_dir(fp)
setup_envir()

rm(list=ls())